### PR TITLE
Add empty folder indicator

### PR DESCRIPTION
**Pull request by PVince81**
## Description
Add empty folder indicator

## Related Issue
Fixes https://github.com/owncloud/phoenix/issues/1910

## How Has This Been Tested?

- test environment:
- test case 1:
- test case 2:
- ...

## Screenshots (if appropriate):
image

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests

## Checklist:

- [x] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:
- [x] improve design, the colors don't match. Bigger size would be nice.
- [x] also process the other lists like "Favorites", "Shared with me", etc
- [x] changelog

### CODE REVIEW

**Comment by PVince81** [comment]
I've added acceptance tests and also tweaked the code that retrieves the list to avoid another potentially waiting code when waiting is not needed => faster results

**Comment by PVince81** [comment]
@individual-it I've pinged you as the tests part is less obvious. I've twisted them a bit to include the "empty message" detection inside the test step that checks if the resource list is empty.

The reason for that is that I wanted to avoid UI-specific language in the Gherkin step like "a message must appear" or something, so I thought having it implied but still checked through assertions would be the way to go. Let me know what you think.

**Comment by LukasHirt** [comment]
> the colors don't match

If you leave out the variation in the icon it will get "system" (default) variation which has the same colour as meta text. Though I'd see this as a workaround only. But until we bring real oC colours into ODS I'd say it's good enough.

File: apps/files/src/components/AllFilesList.vue [hunk 1]
```diff
@@ -37,6 +37,9 @@
               {{ formDateFromNow(item.mdate) }}
             </div>
           </template>
+          <template #emptyMessage>
+            <div class="uk-text-center"><oc-icon :name="$_emptyFolderIcon" type="div" size="large" variation="passive" /><div class="uk-text-muted" v-text="$_emptyFolderText" /></div>
```
Comment (LukasHirt) [inline apps/files/src/components/AllFilesList.vue]: Separate this into own lines pls, it's too long.

File: apps/files/src/components/Collaborators/SharedFilesList.vue [hunk 2]
```diff
@@ -45,6 +45,9 @@
       </div>
       <div class="uk-text-meta uk-text-nowrap uk-width-small" v-text="formDateFromNow(item.shareTime)" />
     </template>
+    <template #emptyMessage>
+      <div class="uk-text-center"><oc-icon name="share" type="div" size="large" variation="passive" /><div class="uk-text-muted" v-text="$_emptyFolderText" /></div>
```
Comment (LukasHirt) [inline apps/files/src/components/Collaborators/SharedFilesList.vue]: Separate this into own lines pls, it's too long.

File: apps/files/src/components/Trashbin.vue [hunk 3]
```diff
@@ -30,6 +30,9 @@
             {{ formDateFromNow(item.deleteTimestamp) }}
           </div>
         </template>
+        <template #emptyMessage>
+          <div class="uk-text-center"><oc-icon name="delete" type="div" size="large" variation="passive" /><div class="uk-text-muted" v-translate>No deleted files</div></div>
```
Comment (LukasHirt) [inline apps/files/src/components/Trashbin.vue]: Separate this into own lines pls, it's too long.

File: apps/files/src/components/AllFilesList.vue [hunk 4]
```diff
@@ -99,6 +102,26 @@ export default {
...
+      if (this.$_isFavoritesList) {
+        return 'star'
+      } else {
+        return 'folder'
+      }
+    },
+
+    $_emptyFolderText () {
+      if (this.$_isFavoritesList) {
+        return this.$gettext('No favorites defined')
```
Comment (LukasHirt) [inline apps/files/src/components/AllFilesList.vue]: Maybe too much technical message? What about something like "No files/folders marked as favorite"?

File: apps/files/src/components/FilesAppBar.vue [hunk 5]
```diff
@@ -50,7 +50,7 @@
             <oc-button v-if="selectedFiles.length > 0" icon="restore" @click="$_ocTrashbin_restoreFiles()">
               <translate>Restore selected</translate>
             </oc-button>
-            <oc-button id="delete-selected-btn" icon="delete" @click="selectedFiles.length < 1 ? $_ocTrashbin_empty() : $_ocTrashbin_deleteSelected()">
+            <oc-button id="delete-selected-btn" icon="delete" :disabled="files.length === 0" @click="selectedFiles.length < 1 ? $_ocTrashbin_empty() : $_ocTrashbin_deleteSelected()">
```
Comment (kulmann) [inline apps/files/src/components/FilesAppBar.vue]: Good catch to disabled the button! However, you can now remove the part from the click handler, that deals with an empty selection.

Comment (PVince81) [reply apps/files/src/components/FilesAppBar.vue]: empty selection doesn't mean no files
if you click the button with no files selected, it's the same as selecting all

**Comment by PVince81** [comment]
I've addressed all comments, except for the design part.

- [ ] We need to update ODS to get the bigger icon sizes: https://github.com/owncloud/owncloud-design-system/pull/632
- [ ] Set size to xlarge and make sure the font size also matches

**Comment by individual-it** [comment]
In the spirit of BDD and the discussions we had in the product repo about how to write steps I would suggest to explicitly have a step that says that there should be the message. I would argue that there is a new behaviour in the application, but that is not reflected in the feature files.
@HanaGemela @phil-davis how do you feel about that?

File: tests/acceptance/stepDefinitions/filesContext.js [hunk 6]
```diff
@@ -320,10 +320,16 @@ When('the user unmarks the favorited file/folder {string} using the webUI', func
...
-  })
+Then('there should be no files/folders/resources listed on the webUI', async function () {
+  const emptyMessageVisible = await client
+    .page
+    .FilesPageElement.filesList()
+    .isEmptyMessageVisible()
+
+  const allRowsResult = await client.page.FilesPageElement.filesList().allFileRows()
+
+  assert.ok(emptyMessageVisible, 'Empty message must be visible')
```
Comment (individual-it) [inline tests/acceptance/stepDefinitions/filesContext.js]: is the message empty, or is the message about the folder being empty? ;-)

Comment (PVince81) [reply tests/acceptance/stepDefinitions/filesContext.js]: the message is about "folder is empty", see screen shot above.

I suspected that this wording will cause confusion, I'll adjust it

**Comment by PVince81** [comment]
> In the spirit of BDD and the discussions we had in the product repo about how to write steps I would suggest to explicitly have a step that says that there should be the message. I would argue that there is a new behaviour in the application, but that is not reflected in the feature files.
> @HanaGemela @phil-davis how do you feel about that?

I was also a bit confused here and went for the idea that the "empty folder" message is an implementation/design detail (which could be considered obvious in any web app)) not related to the requirements. I guess if we say that it's an explicit product requirement to display a message, then we need it as a Gherkin step.

File: apps/files/src/components/AllFilesList.vue [hunk 7]
```diff
@@ -118,9 +118,17 @@ export default {
...
-        return this.$gettext('Folder is empty')
+        return this.$gettext('There are no resources in this folder.')
+      }
+    },
+
+    $_emptyFolderSubText () {
+      if (this.$_isFavoritesList) {
+        return this.$gettext('You can mark some by clicking on the star icon.')
+      } else {
+        return this.$gettext('Drag files and folder here to upload, or browse files from your computer.') // TODO: link
```
Comment (kulmann) [inline apps/files/src/components/AllFilesList.vue]: - `folder` missing an s (plural)
- maybe replace `computer` with `device`, as we're also targeting mobile

File: apps/files/src/components/AllFilesList.vue [hunk 8]
```diff
@@ -118,9 +118,17 @@ export default {
...
+        return this.$gettext('There are no resources marked as favorite.')
       } else {
-        return this.$gettext('Folder is empty')
+        return this.$gettext('There are no resources in this folder.')
+      }
+    },
+
+    $_emptyFolderSubText () {
+      if (this.$_isFavoritesList) {
+        return this.$gettext('You can mark some by clicking on the star icon.')
```
Comment (kulmann) [inline apps/files/src/components/AllFilesList.vue]: ... `star icon in the file list`? As a new user I might not know where to find the star icon.

File: tests/acceptance/stepDefinitions/filesContext.js (hunk 6 above)
Comment (PVince81) [reply tests/acceptance/stepDefinitions/filesContext.js]: renamed to "noContentMessage"

**Comment by PVince81** [comment]
Fixed all "no content" messages.
Renamed EmptyMessage component to NoContentMessage to avoid confusion.
Separate styling for message and call to action.
Adjusted screenshot in original post.

@kulmann please recheck.

File: apps/files/src/components/Collaborators/SharedFilesList.vue [hunk 9]
```diff
@@ -45,6 +45,14 @@
       </div>
       <div class="uk-text-meta uk-text-nowrap uk-width-small" v-text="formDateFromNow(item.shareTime)" />
     </template>
+    <template #noContentMessage>
+      <no-content-message icon="share">
+        <template #message>
+          <span v-if="$_isSharedWithMe"><translate>You are currently not collaborating on other people's resources.</translate></span>
+          <span v-if="!$_isSharedWithMe"><translate>You are currently not collaborating on any of your resources with other people.</translate></span>
```
Comment (kulmann) [inline apps/files/src/components/Collaborators/SharedFilesList.vue]: `v-else`?

**Comment by PVince81** [comment]
For some reason I intuitively tend to avoid `v-else` as it makes the element depend on the previous one, which means I can't easily move around the element with `v-else`. in some cases, like the RecycleScroller, one needs to scroll a lot to find the original condition.

Anyway, I think we can agree to use `v-else` still. I've adjusted accordingly.

File: apps/files/src/components/AllFilesList.vue [hunk 10]
```diff
@@ -37,6 +37,16 @@
               {{ formDateFromNow(item.mdate) }}
             </div>
           </template>
+          <template #noContentMessage>
+            <no-content-message v-if="!$_isFavoritesList" icon="folder">
+              <template #message><translate>There are no resources in this folder.</translate></template>
```
Comment (LukasHirt) [inline apps/files/src/components/AllFilesList.vue]: Why not using the slot directly on `translate` tag?

File: apps/files/src/components/Collaborators/SharedFilesList.vue [hunk 11]
```diff
@@ -45,6 +45,14 @@
       </div>
       <div class="uk-text-meta uk-text-nowrap uk-width-small" v-text="formDateFromNow(item.shareTime)" />
     </template>
+    <template #noContentMessage>
+      <no-content-message icon="share">
+        <template #message>
+          <span v-if="$_isSharedWithMe"><translate>You are currently not collaborating on other people's resources.</translate></span>
```
Comment (LukasHirt) [inline apps/files/src/components/Collaborators/SharedFilesList.vue]: `translate` tags default to span. Could we then either use `v-translate` directly on the span or dropping the span and moving the if statement to the translate tag? I'd prefer the former.

File: apps/files/src/components/AllFilesList.vue (hunk 10 above)
Comment (PVince81) [reply apps/files/src/components/AllFilesList.vue]: Do you mean putting v-translate on the template tag ? I wasn't sure whether this works.
I guess I could test it with a dummy temporary local translation

**Comment by PVince81** [comment]
@LukasHirt I've addressed your comments and moved the `v-translate` on the previous tag.

also fixed the tests

File: apps/files/src/components/AllFilesList.vue (hunk 10 above)
Comment (LukasHirt) [reply apps/files/src/components/AllFilesList.vue]: Not sure if that works. I thought more about `<translate v-slot:message>`

Comment (LukasHirt) [reply apps/files/src/components/AllFilesList.vue]: Ah, now I see this was deprecated in 2.6.0. My bad...

**Comment by PVince81** [comment]
I did a few tests with translations:
- `<template #message v-translate>` the text stays in english
- `<translate v-slot:message>` kills the text completely
- `<span v-translate>` works correctly

**Comment by PVince81** [comment]
@LukasHirt I went with `<span v-translate>` now to avoid span duplication.

**Comment by LukasHirt** [comment]
@PVince81 Thanks for those tests! I saw that the `v-translate` is now in some parts on `<template>`. If I got your test right it's not translating, right? Maybe just a leftover from tests then?

**Comment by PVince81** [comment]
@LukasHirt forgot pushing, please rechech

**Comment by PVince81** [comment]
I messed up the test sentences while doing manual search and replace... Fixed now.

**Comment by PVince81** [comment]
Yeah, the tests fail because they use an existing skeleton folder which contains the file `.gitkeep` :palm_tree: :face_with_head_bandage:

**Comment by PVince81** [comment]
Seems the "no content" check is not compatible with the files drop page, so I've added an URL check and disabled the specific check for the message presence when on the files-drop page.

File: apps/files/src/components/FileList.vue [hunk 12]
```diff
@@ -2,7 +2,7 @@
   <!-- TODO: Take care of outside click overall and not just in files list -->
   <div :id="id" class="uk-height-1-1 uk-position-relative" @click="hideRowActionsDropdown">
     <div class="uk-flex uk-flex-column uk-height-1-1">
-      <oc-grid gutter="small" flex id="files-table-header" class="uk-padding-small">
+      <oc-grid gutter="small" flex id="files-table-header" class="uk-padding-small" v-if="fileData.length > 0" >
```
Comment (LukasHirt) [inline apps/files/src/components/FileList.vue]: Add key pls to `v-if`/`v-else`
